anndata/_core/merge.py

```diff
@@ -240,6 +241,17 @@ def unify_dtypes(dfs: Iterable[pd.DataFrame]) -> list[pd.DataFrame]:
     return dfs
 
 
+def concat_with_unified_dtypes(
+    dfs: Iterable[pd.DataFrame],
+    *,
+    join: Literal["inner", "outer"] = "outer",
+    axis: Literal[0, 1] = 0,
+) -> pd.DataFrame:
+    dfs = _unify_dtypes(dfs)
+    nonempty_dfs = [df for df in dfs if len(df)]
```

Comment:
TODO for me: figure out if this has any effect. Also probably better to filter *before* dtype unification.

But then this can just be:

```python
pd.concat(unify_dtypes(filter(len, dfs)), ...)
```

And doesn't need to be a function.

Reply:
I also have not been able to reproduce this warning so far. Which tests were you getting this from @flying-sheep?

You mentioned: https://github.com/pandas-dev/pandas/blob/2a65fdd227734cbda5d8e33171e857ab4aaeb9d5/pandas/core/internals/concat.py#L488-L500 and `_concatenate_join_units`, but I'm not seeing the connection. I don't immediately see how the referenced code relates to dataframes with no rows.

Reply:
well that’s the code path that gets hit and that’s the warning that gets displayed / converted into an error.

Reply:
Can you show me which tests throw the warning and share your environment info with me?

Reply:
Alright, it's half the `test_concat_size_0_dim` cases where this is coming up. 

I've also done some reading into the warning, and I *like* the new behavior, and think we should use it. With the change in pandas, the `.dtypes` of the output can be figured out from the `.dtypes` of the inputs. To me, this is good and expected. I don't want to opt-out of that when it becomes the default behavior.

So I'm thinking we keep this warning user visible so if it come up, people know where to look. This also makes the behavior change.

I think we can just ignore it during the test, though maybe this could be a `pytest.raises` so we know when to remove it?

Reply:
Would be nice if there was a way to opt in to the new behavior earlier, but I guess leaving it to pandas makes it not our problem.

Reply:
> I think we can just ignore it during the test, though maybe this could be a pytest.raises so we know when to remove it?

OK!